include/lbann/layers/transform/max.hpp

```diff
@@ -0,0 +1,196 @@
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_LAYER_MAX_HPP_INCLUDED
+#define LBANN_LAYER_MAX_HPP_INCLUDED
+
+#include "lbann/layers/transform/transform.hpp"
+
+namespace lbann {
+
+/** Max layer.
+ *  This layer outputs the entrywise maximum of its input tensors.
+ */
+template <data_layout T_layout = data_layout::DATA_PARALLEL, El::Device Dev = El::Device::CPU>
+class max_layer : public transform_layer {
+
+ public:
+  max_layer(lbann_comm *comm,
+            cudnn::cudnn_manager *cudnn = nullptr)
+    : transform_layer(comm) {
+
+    /// @todo Implement
+    static_assert(Dev == El::Device::CPU,
+                  "max layer currently only supports CPU");
+
+    // Max layer has no limit on parents
+    m_expected_num_parent_layers = -1;
+
+  #ifdef LBANN_HAS_CUDNN
+    // Initialize GPU if available
+    this->m_cudnn = cudnn;
+  #endif // LBANN_HAS_CUDNN
+
+  }
+
+  max_layer* copy() const override { return new max_layer(*this); }
+  std::string get_type() const override { return "max"; }
+  data_layout get_data_layout() const override { return T_layout; }
+  El::Device get_device_allocation() const override { return Dev; }
+
+ protected:
+
+  void setup_dims() override {
+    transform_layer::setup_dims();
+    for (const auto& parent : this->m_parent_layers) {
+      const auto& parent_dims = parent->fp_output_dims(this);
+      if (m_neuron_dims != parent_dims) {
+        std::stringstream err;
+        err << "layer " << get_name() << " expects inputs with "
+            << "dimensions ";
+        for (size_t i = 0; i < m_neuron_dims.size(); ++i) {
+          err << (i > 0 ? "x" : "") << m_neuron_dims[i];
+        }
+        err << ", but layer " << parent->get_name() << " outputs with "
+            << "dimensions ";
+        for (size_t i = 0; i < parent_dims.size(); ++i) {
+          err << (i > 0 ? "x" : "") << parent_dims[i];
+        }
+        LBANN_ERROR(err.str());
+      }
+    }
+  }
+
+  void fp_compute() override {
+    const int num_parents = get_num_parents();
+    auto& output = get_activations();
+    switch (num_parents) {
+    case 0: El::Zero(output); break;
```

Comment:
Can we enforce that min and max layers have at least one parent?

Reply:
That sounds reasonable. I'll add that restriction.